### PR TITLE
Data validation changes

### PR DESCRIPTION
**Pull request by yahskapar**
This PR makes two key changes:

1. Each trainer is modified such that code related to the validation dataset and model selection is NOT executed when USE_LAST_EPOCH is set to True.
2. Main.py and config.py are modified to better indicate to users whether or not the validation dataset matters and to catch any weird behavior a user may encounter if they modify their configuration file incorrectly.

I don't think the config (YAML) files themselves have to be updated, especially since any additional functionality related to using a validation dataset (e.g., preprocessing, loading) should only ever occur if a validation dataset is actually being used for model selection in subsequent testing.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -158,53 +164,56 @@ def signal_method_inference(config, data_loader):
...
+                    num_workers=16,
+                    batch_size=config.TRAIN.BATCH_SIZE,
+                    shuffle=True,
+                    worker_init_fn=seed_worker,
+                    generator=g
+                )
+            else:
+                data_loader_dict['train'] = None
+
+        if config.VALID.DATA.DATASET is not None:
```
Comment (yahskapar) [inline main.py:184]: I repeated this pattern for config.TRAIN... and config.TEST... below just to make things look a bit cleaner (and since it makes sense as an additional guard if no dataset type was specified).

Comment (xliucs) [reply main.py:184]: I think we should do `config.TRAIN.DATA.DATASET is not None and config.TRAIN.DATA.DATA_PATH`

File: config.py [hunk 2]
```diff
@@ -255,34 +255,38 @@ def update_config(config, args):
...
-                                                        str(config.VALID.DATA.END) + '.csv')
-    elif ext != '.csv':
-        raise ValueError(self.name, 'FILE_LIST_PATH must either be a directory path or a .csv file name')
-
-    if ext == '.csv' and config.VALID.DATA.DO_PREPROCESS:
-        raise ValueError(self.name, 'User specified FILE_LIST_PATH .csv file already exists. \
-                         Please turn DO_PREPROCESS to False or delete existing FILE_LIST_PATH .csv file.')
+    if config.TEST.USE_LAST_EPOCH is False:
+        if config.VALID.DATA.DATASET is None:
+            raise ValueError(self.name, 'Validation dataset is not provided despite USE_LAST_EPOCH being False!')
```
[the remark this answers is not in the document]
Comment (xliucs) [reply config.py:260]: Can you combine these two if statements? Also, it is preferred to use `not config.TEST.USE_LAST_EPOCH and not config.VALID.DATA.DATASET`

[the remark this answers is not in the document]
Comment (xliucs) [reply config.py:260]: Probaby double check if `config.VALID.DATA.DATASET is None` == `not config.VALID.DATA.DATASET`

File: neural_methods/trainer/DeepPhysTrainer.py [hunk 3]
```diff
@@ -70,15 +72,17 @@ def train(self, data_loader):
...
                 tbar.set_postfix({"loss": loss.item(), "lr": self.optimizer.param_groups[0]["lr"]})
-            valid_loss = self.valid(data_loader)
-            self.save_model(epoch)
-            print('validation loss: ', valid_loss)
-            if (valid_loss < min_valid_loss) or (valid_loss < 0):
-                min_valid_loss = valid_loss
-                self.best_epoch = epoch
-                print("update best model,best epoch :{}".format(self.best_epoch))
+            if self.config.TEST.USE_LAST_EPOCH is False: 
+                valid_loss = self.valid(data_loader)
```
Comment (xliucs) [inline neural_methods/trainer/DeepPhysTrainer.py:76]: Change to not instead of using `== False`

File: neural_methods/trainer/EfficientPhysTrainer.py [hunk 4]
```diff
@@ -79,15 +81,17 @@ def train(self, data_loader):
...
                 self.save_model(epoch)
-        print("best trained epoch:{}, min_val_loss:{}".format(self.best_epoch, min_valid_loss))
+                print('validation loss: ', valid_loss)
+                if (valid_loss < min_valid_loss) or (valid_loss < 0):
+                    min_valid_loss = valid_loss
+                    self.best_epoch = epoch
+                    print("Update best model! Best epoch: {}".format(self.best_epoch))
+                    self.save_model(epoch)
+        if self.config.TEST.USE_LAST_EPOCH is False: 
+            print("best trained epoch:{}, min_val_loss:{}".format(self.best_epoch, min_valid_loss))
```
Comment (xliucs) [inline neural_methods/trainer/EfficientPhysTrainer.py:94]: same

File: neural_methods/trainer/EfficientPhysTrainer.py [hunk 5]
```diff
@@ -79,15 +81,17 @@ def train(self, data_loader):
...
                 tbar.set_postfix(loss=loss.item())
-            valid_loss = self.valid(data_loader)
-            self.save_model(epoch)
-            print('validation loss: ', valid_loss)
-            if (valid_loss < min_valid_loss) or (valid_loss < 0):
-                min_valid_loss = valid_loss
-                self.best_epoch = epoch
-                print("Update best model! Best epoch: {}".format(self.best_epoch))
+            if self.config.TEST.USE_LAST_EPOCH is False: 
+                valid_loss = self.valid(data_loader)
```
Comment (xliucs) [inline neural_methods/trainer/EfficientPhysTrainer.py:85]: same

File: neural_methods/trainer/EfficientPhysTrainer.py [hunk 6]
```diff
@@ -46,7 +46,9 @@ def train(self, data_loader):
         """ TODO:Docstring"""
         if data_loader["train"] is None:
             raise ValueError("No data for train")
-        min_valid_loss = 1
+        if self.config.TEST.USE_LAST_EPOCH is False: 
+            min_valid_loss = 1
```
Comment (xliucs) [inline neural_methods/trainer/EfficientPhysTrainer.py:50]: same

**Review by xliucs — COMMENTED** [review]
Pleae address my comments.

File: config.py [hunk 7]
```diff
@@ -255,34 +255,38 @@ def update_config(config, args):
...
+            config.VALID.DATA.FILE_LIST_PATH = os.path.join(config.VALID.DATA.FILE_LIST_PATH, \
+                                                            config.VALID.DATA.EXP_DATA_NAME + '_' + \
+                                                            str(config.VALID.DATA.BEGIN) + '_' + \
+                                                            str(config.VALID.DATA.END) + '.csv')
+        elif ext != '.csv':
+            raise ValueError(self.name, 'FILE_LIST_PATH must either be a directory path or a .csv file name')
+
+        if ext == '.csv' and config.VALID.DATA.DO_PREPROCESS:
+            raise ValueError(self.name, 'User specified FILE_LIST_PATH .csv file already exists. \
+                            Please turn DO_PREPROCESS to False or delete existing FILE_LIST_PATH .csv file.')
```
[the remark this answers is not in the document]
Comment (yahskapar) [reply config.py:289]: Nope, just indented it and added the if statements immediately prior to this chunk of code.

File: config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yahskapar) [reply config.py]: > Can you replace all the `self.name` to `self.dataset_name`? We have changed that naming convention in base dataloader.

I replaced these with string literals instead since get_config, and subsequently the function to update the config, are only called once and aren't passed the dataset name as is done in the base dataloader.

File: neural_methods/trainer/DeepPhysTrainer.py [hunk 8]
```diff
@@ -42,7 +42,9 @@ def train(self, data_loader):
         """ TODO:Docstring"""
         if data_loader["train"] is None:
             raise ValueError("No data for train")
-        min_valid_loss = 1
+        if self.config.TEST.USE_LAST_EPOCH is False: 
+            min_valid_loss = 1
```
[the remark this answers is not in the document]
Comment (yahskapar) [reply neural_methods/trainer/DeepPhysTrainer.py:46]: This seems to just be an initial value for this chunk of code later on in the train() function:

```
                if (valid_loss < min_valid_loss) or (valid_loss < 0):
                    min_valid_loss = valid_loss
                    self.best_epoch = epoch
                    print("Update best model! Best epoch: {}".format(self.best_epoch))
                    self.save_model(epoch)
```

My current understanding is as follows:

- For TS-CAN, EfficientPhys, and DeepPhys, which all use MSE loss, the first term in the if statement is utilized to set the initial min_valid_loss value. If it's less than the initialized value of 1, it's set to to the new value and a new best epoch is found. The MSE loss can never be negative, so the second term should never be triggered. The assumption here seems to be that the valid loss has to get below 1 to be updated as the best epoch for the first time. @xliucs, maybe this can be changed such that `current_valid_loss` is used, set to `None` to begin with, and when updated for the first time we check if's `None`, set it if so, and use it moving forward as the min_valid_loss.
- For PhysNet, which uses a negative Pearson correlation loss, the second term (valid_loss < 0) is triggered to set the initial min_valid_loss. Subsequently, as the negative Pearson correlation loss becomes more negative, the best epoch is updated.

[the remark this answers is not in the document]
Comment (yahskapar) [reply neural_methods/trainer/DeepPhysTrainer.py]: I just updated this PR with a commit (0b0b8d4dfaeb52b3897c559530abcf6e8b784570) that I believe makes more sense. Let me know what you think, @xliucs.

File: main.py [hunk 9]
```diff
@@ -142,9 +142,15 @@ def signal_method_inference(config, data_loader):
             valid_loader = data_loader.PURELoader.PURELoader
         elif config.VALID.DATA.DATASET == "SCAMPS":
             valid_loader = data_loader.SCAMPSLoader.SCAMPSLoader
+        elif config.VALID.DATA.DATASET is None:
+            if config.TEST.USE_LAST_EPOCH is False:
+                raise ValueError("Validation dataset not specified despite USE_LAST_EPOCH set to False!")
```
[the remark this answers is not in the document]
Comment (xliucs) [reply main.py:147]: @yahskapar  Can you address this one?

**Review by xliucs — CHANGES_REQUESTED** [review]

File: main.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yahskapar) [reply main.py]: Done.

**Review by xliucs — APPROVED** [review]
LGTM
